src,tools: initialize cppgc

dharesign opened the pull request:
This patch:

- Initializes cppgc in InitializeOncePerProcess() when
  kNoInitializeCppgc is not set
- Create a CppHeap and attach it to the Isolate when
  there isn't one already during IsolateData initialization.
  The CppHeap is detached and terminated when IsolateData
  is freed.
- Publishes the cppgc headers in the tarball.

This allows C++ addons to start using cppgc to manage objects.

A helper node::SetCppgcReference() is also added to help addons
enable cppgc tracing in a user-defined object.

Refs: https://github.com/nodejs/node/issues/40786
Refs: https://docs.google.com/document/d/1ny2Qz_EsUnXGKJRGxoA-FXIE2xpLgaMAN6jD7eAkqFQ/edit

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

dharesign commented on src/node.cc:
```diff
@@ -1046,6 +1046,15 @@ InitializeOncePerProcessInternal(const std::vector<std::string>& args,
   }
 
+  if (!(flags & ProcessInitializationFlags::kNoInitializeCppgc)) {
+    // V8 falls back to an internal page_allocator if one isn't specified.
```
The `page_allocator` typically comes from the `v8::Platform`, which only exists if `kNoInitializeNodeV8Platform` is not specified, however the `node::Platform` isn't given one at construction, so in reality it's always `nullptr`.  It used to be required, but v8 fixed that in 10.9.166 (https://github.com/v8/v8/commit/b28d72b5efe4e2969dee9f0b850744cad30f561d).

joyeecheung replied:
Since `page_allocator` is always nullptr here, we might as well just be explicit and  pass nullptr, and mention that unless `kNoInitializeCppgc` is set, we always end up using the internal page allocator (and maybe DCHECK that `!(flags & ProcessInitializationFlags::kNoInitializeNodeV8Platform)` implies `per_process::v8_platform.Platform()->GetPageAllocator()` should that change one day).

dharesign replied:
I assume you mean this?
```
DCHECK_IMPLIES(
    !(flags & ProcessInitializationFlags::kNoInitializeNodeV8Platform),
    !per_process::v8_platform.Platform()->GetPageAllocator());
```

aka: if we are initializing the platform, it doesn't have a page allocator that we want to pass along.

dharesign replied:
I added the above `DCHECK_IMPLIES`.  Let me know if I misunderstood.

joyeecheung replied:
Yes, that's what I meant, thanks

dharesign commented on src/node_main_instance.cc:
```diff
@@ -84,6 +88,17 @@ NodeMainInstance::NodeMainInstance(const SnapshotData* snapshot_data,
+          WrapperDescriptor(
+              BaseObject::kEmbedderType,
+              BaseObject::kSlot,
+              0x90df
```
A recent commit made `kEmbedderType` and `kSlot` the correct values here (https://github.com/nodejs/node/commit/8b0e5b19bd0).  The embedder id in that commit was specified as `0x90de`.  When Chromium initializes the `CppHeap` its embedder id is `0x1` (`kEmbedderBlink`).  Pre-cppgc it used to be the case that for embedder tracing you had to specify the embedder id when you create objects, and it would need to match.  With `cppgc`, you just use `cppgc::MakeGarbageCollected<>` and I believe it uses the values that were specified when the `CppHeap` was created.  As a result, add-ons using `cppgc` don't need to be aware of the value chosen by Node.  Given though that the reason `0x90de` was picked was so that non-`cppgc` objects don't get mixed up with `cppgc` objects, we probably want to use a different embedder id than `0x90de` which Node is currently using for non-`cppgc` objects.  I just picked the next number here.

joyeecheung replied:
I've rebased this on top of https://github.com/nodejs/node/pull/48660 so we do not need to worry about the exact values of the IDs anymore.

dharesign commented on tools/install.py:
```diff
@@ -197,15 +197,56 @@ def files(action):
 def headers(action):
   def wanted_v8_headers(files_arg, dest):
     v8_headers = [
+      'deps/v8/include/cppgc/allocation.h',
```
It might be better to go back to excluding the headers you don't want to export?

Whilst I'm here though, I'll mention that we actually do make use of the `v8-inspector.h` and `v8-inspector-protocol.h` headers, so we have a local patch to include those also.  Those were the only headers that used to be in the disallow list (See: https://github.com/nodejs/node/commit/38f32386c138073c6a020ce79085daea15e7b800).

joyeecheung replied:
I think the concern is that we don't want to accidentally expand the API surface. If we have a disallow list we might fail to notice when V8 adds new headers and then breaks the API by removing them. If the list is explicit it helps us notice about changes like this and we can make some fixups or adjust the timing of the V8 upgrade.

targos replied:
I don't understand the subject enough to have an opinion on whether we should expose all these additional headers, but I'm afraid that doing so will effectively block us from doing semver-minor V8 upgrades in the future.

dharesign replied:
FYI, The v8 headers missing from this list (as of 11.3) are:

```
+      'deps/v8/include/v8-fast-api-calls.h',
+      'deps/v8/include/v8-inspector-protocol.h',
+      'deps/v8/include/v8-inspector.h',
+      'deps/v8/include/v8-metrics.h',
+      'deps/v8/include/v8-unwinder-state.h',
+      'deps/v8/include/v8-util.h',
+      'deps/v8/include/v8-value-serializer-version.h',
+      'deps/v8/include/v8-version-string.h',
+      'deps/v8/include/v8-wasm-trap-handler-posix.h',
+      'deps/v8/include/v8-wasm-trap-handler-win.h',
```

targos replied:
Assuming we publish all these internal cppgc headers. Can we still update V8 in an ABI-compatible way if they make breaking changes to them (while keeping `v8-cppgc.h` without breaking changes)

dharesign replied:
`v8-cppgc.h` does not provide everything you need.  You need to directly include a bunch of the `cppgc/` headers directly.  In our existing code, we are including the following:

```
$ git grep "#include.*cppgc" | cut -d ":" -f2 | sort -u
#include <cppgc/allocation.h>
#include <cppgc/garbage-collected.h>
#include <cppgc/name-provider.h>
#include <cppgc/persistent.h>
#include <cppgc/prefinalizer.h>
#include <v8-cppgc.h>
```

The internal headers I assume you're not supposed to directly include, but they're included by the non-internal headers, so they need to be available for transitive inclusion.

You would have to be as careful as you are now about changes to these headers, when determining whether it's safe to do a given v8 upgrade as a node minor release.

joyeecheung replied:
From what I can tell, if cppgc's ABI compatibility is included in our ABI compatibility contract (no breaking changes in semver-minor), we have more changes to review when doing a V8 upgrade. Also because how the cppgc constructs are template-heavy, it's easier for the cppgc updates to break the ABI.

On the other hand, I think we could also consider excluding cppgc from our ABI compatibility contract, or at least list them as experimental so they are excluded from the contract for now until the update to the headers really slow down. Our ABI compatibility contract only really extends to V8, whereas cppgc is more like a standalone C++ GC library that happens to be bundled within V8. Before this users would not be able to use cppgc in their addons anyway so we could state in the beginning that if the addons use cppgc, we do not guarantee that the addons can be loaded in same major releases. For the vast majority of addons that do not use cppgc it would not matter because the new headers are just added symbols which are ABI compatible. We should be able to enforce this by making the cppgc stuff only visible when the addons explicitly request it (e.g. adding an ifdef to the headers, so the headers are still part of the tarball but addons now need to get access to what's in cppgc via an explicit define, and when they do that they are opting into being at their own risk in terms of ABI compatibility)

joyeecheung replied:
Actually I took a look at the 11.3.244.8...11.6.130 changes (V8 on the main branch here and the latest V8 release) and it seems there aren't actually that many ABI breakages in cppgc - most changes are additions - compared to what happens in other V8 headers. If we maintain the current practice, which is upgrading V8 a few times (no more than once per month, I gather) before the release goes LTS, cppgc probably would not be that much of a problem (other than we would have a bit more changes to review). We do still need to pay a bit more attention to the template changes, but this also already happens to the fast API calls header (also template-heavy) and there is also growing number of inlined implementations in other headers that lead to similar problems. cppgc probably only contributes <10% to the overall ABI compat problem.

> Assuming we publish all these internal cppgc headers. Can we still update V8 in an ABI-compatible way if they make breaking changes to them (while keeping v8-cppgc.h without breaking changes)?

From what I can tell if the users do not actually use any thing from `cppgc/` it does not matter for them. And they are not supposed to use anything from `cppgc/internal` (V8 very explicitly warns against using anything there) so changes to the internal headers would matter less. The safer choice might still be that we state that we do not guarantee ABI compatibility for `cppgc` for now but might consider including it when we think it's ready.

joyeecheung replied:
Opened https://bugs.chromium.org/p/v8/issues/detail?id=14062 in the upstream about cppgc ABI compatibility.

joyeecheung replied:
From what I got from the v8 issue above my conclusion is that the cppgc headers should not introduce that much ABI breakages for the 6-month cadences we have, it's also only a concern for users who actually use any of the cppgc APIs, which should be niche at least for now. On the safe side we can also just state that embedder support for cppgc is experimental and it is not included in our ABI compatibility guarantee for now.

dharesign commented on src/node_main_instance.cc:
```diff
@@ -84,6 +88,17 @@ NodeMainInstance::NodeMainInstance(const SnapshotData* snapshot_data,
   SetIsolateCreateParamsForNode(isolate_params_.get());
   Isolate::Initialize(isolate_, *isolate_params_);
 
+  cpp_heap_ = CppHeap::Create(
```
Not sure if this should be optional, or conditional on the earlier `kNoInitializeCppgc`?

joyeecheung replied:
Can you add a test for `kNoInitializeCppgc`, similar to the one in `embedtest.cc`? I assume doing `CppHeap::Create()` and `Isolate:: AttachCppHeap()` right after `CommonEnvironmentSetup::Create()` would be timely enough? It would be good to know when we add something in `CommonEnvironmentSetup::Create()` that makes it no longer the case though.

I think for `NodeMainInstance()` we can probably just do this unconditionally for the add-ons. It does not seem to be necessary for embedders who use `NodeMainInstance()`  and also enable cppgc for add-ons (do they exist?) to customize this part.

dharesign replied:
I can add `kNoInitializeCppgc` to `embedtest.cc` and have it call `cppgc::InitializeProcess` manually.  `CppHeap::Create()` and `Isolate::AttachCppHeap()` happen internally in `CommonEnvironmentSetup` however.  Are you saying you don't want that to be the case?

I think it makes sense, if Node starts allowing embedders to use `cppgc`, to expect that embedders specifying `kNoInitializeCppgc` need to manually initialize it, similar to how they are currently required to initialize v8.

addaleax replied:
> for embedders who use `NodeMainInstance()` and also enable cppgc for add-ons (do they exist?)

I wouldn’t spend too much time worrying about embedders who don’t stick to the public API :slightly_smiling_face:

joyeecheung replied:
Oh, I misread, there is only one flag for the process initialization, so it's not propagated to environments. So I guess we can just add a flag for cppgc creation to `IsolateSettingsFlags`, and then do the calls unconditionally in `CommonEnvironmentSetup::Create()` (this means that either the user cannot use `CommonEnvironmentSetup` if they want to configure the cppgc heap, but that's probably fine).

dharesign commented:
In Electron, the cppgc heap will be initialized by Chromium.  I don't think anything in this PR clashes with what Electron will want.

When they initialize node, they can use `kNoInitializeCppgc` here:
https://github.com/electron/electron/blob/4e661842879ae0cb01a9fc251d56c8cabc958466/shell/app/node_main.cc#L161-L165

They don't use `NodeMainInstance` or `CommonEnvironmentSetup`.

CC: @codebytere / @nornagon for awareness

dharesign commented on src/api/embed_helpers.cc:
```diff
@@ -107,6 +112,17 @@ CommonEnvironmentSetup::CommonEnvironmentSetup(
+          WrapperDescriptor(
+              BaseObject::kEmbedderType,
+              BaseObject::kSlot,
+              0x90df)
```
I don't think the exact value matters a whole lot.  In Electron, Chromium will initialize the heap with `0x1`.  Node addons making use of `cppgc` don't need to specify the value, it's automatically set by `cppgc::MakeGarbageCollected`.

What does matter, is that it doesn't match the value used in the embedder fields for non-cppgc objects.  That was why https://github.com/nodejs/node/commit/8b0e5b19bd0 updated the value in the first place.

We could certainly expose a method that uses this fixed value that embedders can use.

dharesign replied:
Question is: where?  It could potentially be something extra done in `SetIsolateUpForNode` (that _is_ called by Electron, so I guess it would need to be optionally done based on something in `IsolateSettings`).  Or it could be a new method.

addaleax replied:
> that _is_ called by Electron, so I guess it would need to be optionally done based on something in `IsolateSettings`

I think that would seem okay to me :+1: `IsolateSettings` was added specifically so that Electron can customize this a bit more iirc.

dharesign replied:
One issue with this is that the `CppHeap` needs to be owned by someone.  With the changes I have thus far, it's either owned by `CommonEnvironmentSetup::Impl`, or `NodeMainInstance`.

`SetIsolateUpForNode` doesn't have any place to own the `CppHeap`.

addaleax replied:
I guess we could make it part of the `MultiIsolatePlatform`’s job to attach/detach the cpp heap in `RegisterIsolate`/`UnregisterIsolate`? And if we need this behavior to be toggle-able in the default implementation, we could add an extra argument to `MultiIsolatePlatform::Create()`…

joyeecheung replied:
So far it doesn't seem necessary to toggle this in `NodePlatform`. Also AFAICT, where Electron would need us to skip the cppgc initialization it should also configure `ProcessInitializationFlags::kNoInitializeNodeV8Platform` too? So I am thinking if we can just use that flag to decide whether we need to do cppgc initialization, as in the `NodePlatform` we also have a reference to the page allocator that can be used to initialize cppgc, and this way we can put the cppgc-related registration/initialization cold all in the NodePlatform implementation if we also attach/detach the cpp heap in `RegisterIsolate/UnregisterIsolate`.

joyeecheung replied:
On a closer look, `IsolateData` might be a better place to own the CppHeap. And it should be safe to attach/detach in its constructor/destructor.

joyeecheung replied:
Apparently our own tests don't actually treat `IsolateData` as per-isolate even though they are meant to be that way...I've sent https://github.com/nodejs/node/pull/48450 to address that. If that doesn't work, we could still go with RegisterIsolate/UnregisterIsolate.

joyeecheung replied:
Moved the CppHeap to IsolateData, which now detects if the cpp heap is already created before creating a cpp heap.

dharesign commented:
Erp, I hit the close button by mistake.  Please re-approve running the workflows >_<

dharesign commented:
The previous test runs seemed to fail, and actually running my local build of node likewise fails.  It seems to be due to a bug in v8:

https://github.com/v8/v8/blob/main/src/heap/cppgc/heap-base.cc#L104

I believe this should be `delgate_` not `delegate`: the latter is moved from.

CC: @mlippautz

dharesign commented:
> The previous test runs seemed to fail, and actually running my local build of node likewise fails. It seems to be due to a bug in v8:
> 
> https://github.com/v8/v8/blob/main/src/heap/cppgc/heap-base.cc#L104
> 
> I believe this should be `delgate_` not `delegate`: the latter is moved from.
> 
> CC: @mlippautz

Yep, changing that fixes my local build.

dharesign commented on src/api/embed_helpers.cc:
```diff
@@ -107,6 +112,14 @@ CommonEnvironmentSetup::CommonEnvironmentSetup(
+      CppHeapCreateParams{
+          {},
+          WrapperDescriptor(
+              BaseObject::kEmbedderType, BaseObject::kSlot, 0x90df)});
```
I added `kNodeEmbedderIdForCppgc`.

dharesign commented:
> > The previous test runs seemed to fail, and actually running my local build of node likewise fails. It seems to be due to a bug in v8:
> > https://github.com/v8/v8/blob/main/src/heap/cppgc/heap-base.cc#L104
> > I believe this should be `delgate_` not `delegate`: the latter is moved from.
> > CC: @mlippautz
> 
> Yep, changing that fixes my local build.

I added a second commit to this PR which fixes the above.  We can switch the commit to cherry-pick the v8 commit once it gets fixed there.  Hopefully the CI checks will proceed now.

dharesign commented:
Node is still segfaulting when I run it locally:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x0000000100a5c14c node`cppgc::internal::MarkerBase::EnterAtomicPause(cppgc::EmbedderStackState) [inlined] cppgc::internal::SingleThreadedHandle::Cancel(this=<unavailable>) at task-handle.h:27:20 [opt]
   24
   25  	  void Cancel() {
   26  	    DCHECK(is_cancelled_);
-> 27  	    *is_cancelled_ = true;
   28  	  }
   29
   30  	  bool IsCanceled() const {
Target 0: (node) stopped.
(lldb) frame sel 1
frame #1: 0x0000000100a5c144 node`cppgc::internal::MarkerBase::EnterAtomicPause(this=0x0000000106808200, stack_state=kNoHeapPointers) at marker.cc:260:33 [opt]
   257 	    // Cancel remaining incremental tasks. Concurrent marking jobs are left to
   258 	    // run in parallel with the atomic pause until the mutator thread runs out
   259 	    // of work.
-> 260 	    incremental_marking_handle_.Cancel();
   261 	    heap().stats_collector()->UnregisterObserver(
   262 	        incremental_marking_allocation_observer_.get());
   263 	    incremental_marking_allocation_observer_.reset();
(lldb) print incremental_marking_handle_
(cppgc::internal::MarkerBase::IncrementalMarkingTaskHandle) $0 = {
  is_cancelled_ = nullptr {
    __ptr_ = 0x0000000000000000
  }
}
```

It looks like `foreground_task_runner_` is `nullptr`, so `MarkerBase::ScheduleIncrementalTaskRunner` exits early before initializing `incremental_marking_handle_`.  Later, `MarkerBase::EnterAtomicPause` assumes that `incremental_marking_handle_` is set such that it can call `.Cancel()`, but it's not.

dharesign commented:
Submitted the following v8 bugs:
* https://bugs.chromium.org/p/v8/issues/detail?id=13589
* https://bugs.chromium.org/p/v8/issues/detail?id=13590

dharesign commented:
I added a commit to add `SetIsolate` as a virtual method on the `cppgc::Platform` protocol.  Not sure if that's the direction v8 will go to fix the issue, but it should help get us past that issue and check the rest of the CI passes.

dharesign commented:
Please re-approve running the workflows 😅

dharesign commented:
OK I rebased and fixed some lint issues.

There was an outstanding question about what to do regarding the creation and attaching of the cpp heap.  Currently it's being done in either the `NodeMainInstance` or `CommonEnvironmentSetup::Impl`.  There was a suggestion to move it into a common method, but given the `CppHeap` needs to be owned, I don't see a good place to put it.

I think the current PR is a reasonable starting point.  If an embedder wants to create a `CppHeap`, then they can either use their own id, or we can at that time figure out where to put this utility method (or perhaps just a way to get the id to use).

This PR does still have two commits which belong upstream in v8.  One of them has landed in v8 11.1.119 (https://github.com/v8/v8/commit/22ef44b6553d8207aac5766c5fd7624a1e41fbc4).  The other hasn't been addressed yet.  I'm guessing these will want to be cherry-picked / applied outside of this PR?  I have them here so the CI can run successfully.

dharesign commented:
Looks like all CI passed apart from another lint issue.  Will push a fix for that now.

jasnell commented on src/node_main_instance.h:
```diff
@@ -83,6 +83,7 @@ class NodeMainInstance {
   std::unique_ptr<ArrayBufferAllocator> array_buffer_allocator_;
   v8::Isolate* isolate_;
   MultiIsolatePlatform* platform_;
+  std::unique_ptr<v8::CppHeap> cpp_heap_;
```
node_worker.h will likely need this also. Specifically, `v8::CppHeap` can only be used by a single thread at a time, and garbage collected objects managed by the heap are only safe to be cleaned up on the thread currently owning the lock for the heap. To be safe, we probably want each worker thread to have its own `v8::CppHeap` instance associated with its own `v8::Isolate`, and have some mechanism in place to ensure that oilpan gc'd objects are safely being handled cross-thread.

dharesign replied:
Added to `node_worker.cc`: https://github.com/nodejs/node/pull/45704/files#diff-b06b65a7f7e33816a1986d731765d732ab23f585bece6b23c5e83ca3682dba62

Regarding
> and have some mechanism in place to ensure that oilpan gc'd objects are safely being handled cross-thread

If the worker threads have an entirely separate `v8::Isolate`, then there's no cross-thread object sharing that can take place, and we don't need to do anything?

I haven't used worker threads, so I could be wrong, but taking a brief scan of the docs it looks like the object passing is through `ArrayBuffer`, either by transferring ownership of the underlying native memory (in which case the local object becomes empty), or by cloning the underlying native memory (if marked untransferable).  In neither case are the v8 objects being transferred/shared, just the backing data?

jasnell replied:
We have cases where we share underlying objects across multiple threads. For instance, cloned key object data, or Blob internal storage, etc. We basically just need to make sure those type things aren't GC managed, and keep the GC managed things local to their owning threads.

joyeecheung replied:
For cppgc objects the objects are managed by V8, I don't think we need to do anything else. We only need to make sure that the `CppHeap` is alive while the isolate is alive, and that's why we'd need to own the pointer. But the reference to the `CppHeap` should be retrieved via `isolate->GetCppHeap()` instead, and before that the user should've already ensured that they are accessing `isolate` on the right thread.

> We basically just need to make sure those type things aren't GC managed, and keep the GC managed things local to their owning threads.

We don't have any internal objects that are GC managed for now. That would need to be explicitly implemented.

joyeecheung commented:
> I'm guessing these will want to be cherry-picked / applied outside of this PR? I have them here so the CI can run successfully.

Have you uploaded CLs for these to V8 yet?  In general we need the fix to land in upstream first, and then backport them to the main branch. If a PR depends on the fix, either we open a backport PR for that fix and then rebase the Node.js PR once that lands in `deps/v8`, or include the backport commit in the Node.js PR (personally I prefer the first because then it's easier to revert in case something goes wrong), see: https://github.com/nodejs/node/blob/main/doc/contributing/maintaining-V8.md#backporting-with-git-node-recommended

joyeecheung commented on src/node_main_instance.cc:
```diff
@@ -80,6 +84,14 @@ NodeMainInstance::NodeMainInstance(const SnapshotData* snapshot_data,
+                          WrapperDescriptor(BaseObject::kEmbedderType,
+                                            BaseObject::kSlot,
+                                            kNodeEmbedderIdForCppgc)});
+  isolate_->AttachCppHeap(cpp_heap_.get());
```
From the doc looks like this can only be done once per isolate. @codebytere would doing this unconditionally be in conflict with what Electron does? (From https://github.com/nodejs/node/pull/43521 it seems that this would be called by someone else in Electron too).

joyeecheung replied:
This PR only touches `CommonEnvironmentSetup`, `NodeMainInstance` and `WorkerThreadData`, so if Electron isn't using any of these (from what I can tell it's using `CreateEnvironment()` and `LoadEnvironment()` now which should be unaffected by this PR) I think it should be fine.

dharesign commented:
> > I'm guessing these will want to be cherry-picked / applied outside of this PR? I have them here so the CI can run successfully.
> 
> Have you uploaded CLs for these to V8 yet? In general we need the fix to land in upstream first, and then backport them to the main branch. If a PR depends on the V8 fix, either we open a backport PR for that fix and then rebase the Node.js PR once that lands in `deps/v8`, or include the backport commit in the Node.js PR (personally I prefer the first because then it's easier to revert in case something goes wrong, but I guess for this PR it probably doesn't matter because we don't support cppgc before this anyway so the V8 fix could not affect us otherwise), see: https://github.com/nodejs/node/blob/main/doc/contributing/maintaining-V8.md#backporting-with-git-node-recommended

CLs no.  I created issues:
* Use-After-Move DCHECK:
  * https://bugs.chromium.org/p/v8/issues/detail?id=13589
  * Status: Fixed
    * https://github.com/v8/v8/commit/22ef44b6553d8207aac5766c5fd7624a1e41fbc4
  * Version: 11.1.119
* Invalid `static_cast`:
  * https://bugs.chromium.org/p/v8/issues/detail?id=13590
  * Status: Open

dharesign commented:
I pushed a new commit for "[deps: V8 Fix Invalid static_assert](https://github.com/nodejs/node/pull/45704/commits/6987a55a616951ad5efd8e7e8d0089b7894ff46c)" which should address the concern they have with the original patch I suggested.  Hopefully we can then get it landed upstream.

dharesign commented:
OK, I submitted a CL for the invalid `static_cast` issue and it just landed: https://github.com/v8/v8/commit/eea8a2f22bcf2bcfd12a093a0e1ba762b8d0a5ec

I can create a new PR to backport these two fixes now they've both landed in v8 then we can rebase this PR once that lands.

dharesign commented:
This PR is rebased on top of #47065 to cherry pick the required fixes in v8.  Once that PR lands I can rebase this PR on top of the merge, or they can be merged from this PR and we can close #47065.  Whatever works!

dharesign commented:
Rebased on top of `main` which now includes v8 11.3.  That covers the commits we needed to land in v8 for this.

joyeecheung commented:
Is this ready for re-review? Or is it still blocked on anything?

dharesign commented:
> Is this ready for re-review? Or is it still blocked on anything?

No blockers remain! Thanks

joyeecheung commented on src/node_main_instance.cc:
```diff
@@ -49,6 +53,14 @@ NodeMainInstance::NodeMainInstance(const SnapshotData* snapshot_data,
+  cpp_heap_ = CppHeap::Create(
+      platform,
+      CppHeapCreateParams{{},
+                          WrapperDescriptor(BaseObject::kEmbedderType,
```
The type and index is meant for identifying embedder (addons in this case) objects, which I think technically are separate from the internal Baseobjects? I think it would be more robust if we make the convention for addon cppgc wrappers separate - for example, simply declare a `constexpr WrapperDescriptor` in `node.h` with these parameters. (I wonder how users can create cppgc-managed objects without access to the embedder id - do they call `SetAlignedPointerInInternalField()` to some pointer that points to a hard-coded copy of that id?)

joyeecheung replied:
If we are using `BaseObject::kEmbedderType` and `BaseObject::kSlot` here to ensure that the BaseObject layout is not in conflict with this, perhaps we could declare `constexpr WrapperDescriptor` in `node.h`, and internally we `static_assert` that the indices are compatible (matching BaseObject offsets).

dharesign replied:
There can (at least currently) be only one `CppHeap` for a given `isolate`.  All objects that live in that heap need to then all agree upon the ids and slots used here.

If there are two add-ons both using `cppgc` to create objects, they need to be compatible.

The API used to create such an object is:
```cpp
T* obj = cppgc::MakeGarbageCollected<T>(
  isolate->GetCppHeap()->GetAllocationHandle(),
  args...);
```

The `AllocationHandle` encapsulates which `CppHeap` the object is being created in, and therefore which values are used for the ids and slots.  Add-ons do not need to be explicitly aware which ids and slots are being used.

joyeecheung replied:
About https://github.com/nodejs/node/pull/45704#discussion_r1037419323, IIUC this is something that the user needs to do to create cppgc-managed objects in addons:

```c++
uint16_t wrappable_type = kNodeEmbedderIdForCppgc;
// The lifetime of wrappable_object is managed by V8
auto* wrappable_object =
    cppgc::MakeGarbageCollected<Wrappable>(isolate()->GetCppHeap()->GetAllocationHandle());
Local<Object> object = Object::New(isolate)
object->SetAlignedPointerInInternalField(kNodeEmbedderWrappableTypeIndex,
                                         &wrappable_type);
object->SetAlignedPointerInInternalField(kNodeEmbedderWrappableInstanceIndex,
                                         wrappable_object);
```

This is different from Electron's case because there they are configuring the cpp heap, so they are already aware of the wrapper descriptor. But here we are the one picking the wrapper descriptor, so we should actually expose the ID and the `wrappable_instance_index` that we pick, because addons are supposed to use that index to store their `GarbageCollected` wrappables.

dharesign replied:
I guess it depends what you're doing, but if you want to introduce an edge to this `cppgc` object from a `v8::Object` you can put the pointer into the internal field like that, yes.  We do actually do that.

I guess today people use `node::ObjectWrap`.  Maybe there should be a variant of that for `cppgc` that encapsulates all these gnarly details?

joyeecheung replied:
I think this model probably is a bit too different from what `ObjectWrap()` is doing now (which still relies on a persistent handle). Perhaps exposing a helper for setting this is already enough? Something like

```
void SetCppgcWrappable(Local<Object> object, void* garbageCollected);
```

which does the snippet above.

dharesign commented on src/node.h:
```diff
@@ -1476,6 +1478,15 @@ void RegisterSignalHandler(int signal,
 #endif  // _WIN32
 
+// Configure the layout of the JavaScript object with a cppgc::GarbageCollected
+// instance so that when the JavaScript object becomes unreachable, the garbage
```
```suggestion
// instance so that when the JavaScript object is reachable, the garbage
```

joyeecheung commented on src/node.h:
```diff
@@ -1476,6 +1478,15 @@ void RegisterSignalHandler(int signal,
+// for add-ons, but not necessarily for embedders).
+// The object must be created with at least two internal fields available,
+// and the first two internal fields would be configured by Node.js.
+void TraceWithCppGC(v8::Local<v8::Object> js_object, void* garbage_collected);
```
Actually we might not need this if https://chromium-review.googlesource.com/c/v8/v8/+/4598833 lands in the upstream.

joyeecheung commented on tools/install.py:
```diff
@@ -197,15 +197,56 @@ def files(action):
+      'deps/v8/include/cppgc/heap-state.h',
+      'deps/v8/include/cppgc/heap-statistics.h',
+      'deps/v8/include/cppgc/heap.h',
+      'deps/v8/include/cppgc/internal/api-constants.h',
+      'deps/v8/include/cppgc/internal/atomic-entry-flag.h',
+      'deps/v8/include/cppgc/internal/base-page-handle.h',
+      'deps/v8/include/cppgc/internal/caged-heap-local-data.h',
+      'deps/v8/include/cppgc/internal/caged-heap.h',
+      'deps/v8/include/cppgc/internal/compiler-specific.h',
+      'deps/v8/include/cppgc/internal/finalizer-trait.h',
+      'deps/v8/include/cppgc/internal/gc-info.h',
+      'deps/v8/include/cppgc/internal/logging.h',
+      'deps/v8/include/cppgc/internal/member-storage.h',
+      'deps/v8/include/cppgc/internal/name-trait.h',
+      'deps/v8/include/cppgc/internal/persistent-node.h',
+      'deps/v8/include/cppgc/internal/pointer-policies.h',
+      'deps/v8/include/cppgc/internal/write-barrier.h',
```
```suggestion
```

We don't need to expose these

joyeecheung replied:
Oh actually we do, they are needed by the public headers..

joyeecheung commented on src/node.h:
```diff
@@ -1486,6 +1490,23 @@ void RegisterSignalHandler(int signal,
+// If the CppHeap is owned by Node.js, which is usually the case for addon,
+// the object must be created with at least two internal fields available,
+// and the first two internal fields would be configured by Node.js.
+// This helper is only available in a Node.js context.
```
Oh I actually realized that we don't need the access to a full isolate data for this...so we can just implement this as isolate-specific.

joyeecheung commented:
I've rebased this on top of https://github.com/nodejs/node/pull/48660 so that we no longer need to worry about the wrapper ID. The CppHeap is now owned by the IsolateData if it's not already initialized (in the case of Blink's unsanboxed renderers, it's already initialized, then Node.js would recognize it and use the wrapper ID of that instead).

Regarding ABI compatibility, I think we can mark cppgc ABI compatibility as experimental for now, though so far the public part of the cppgc API already has a quite stable ABI compared to other V8 headers.

Regarding thread safety, since we are not creating any internal cppgc-managed objects, this is not yet our concern. Any internal object must explicitly implement the `GarbageCollected` interface to be cppgc-managed, so we could figure that out on a case-by-case basis if necessary when we actually start to migrate internal objects to the CppHeap - it's also fine to leave them the way they are (i.e. allocated outside the CppHeap), as all other memory management utils that we use (e.g. `v8::Persistent`) still work just fine with cppgc, so objects managed differently can live together alright - I believe this is also still the reality in Blink anyway. But at least this PR allows addons to create their own cppgc-managed objects.

I think this PR has no other outstanding questions for now other than that it needs https://github.com/nodejs/node/pull/48660.

joyeecheung commented:
@jasnell @addaleax @targos  @nodejs/startup @nodejs/cpp-reviewers I updated the PR quite a bit after getting consent from @dharesign, so I am not the right person to give approval on this anymore. Can we get some new reviews please? Thanks!

RafaelGSS reviewed (commented):
Shouldn't we have docs for that?

Qard reviewed (approved):
LGTM.

@RafaelGSS I think it would be best to leave it undocumented until we have some of our own internal types using it first to verify the pattern is working the way we want it to before making it a public and supported thing. 🤔

joyeecheung commented:
I've put the link to the design doc in the OP which should be picked up by node-core-utils when this lands.

Still need to wait until the security embargo is lifted before we can launch another CI on it.

UlisesGascon commented:
This PR didn't land cleanly because of https://github.com/nodejs/node/pull/48660 in Node 20.x. Can you backport it?

glaubitz commented:
This change broke NodeJS on big-endian ppc64 on Linux for me.

joyeecheung commented:
I don't think V8 tests for big-endian ppc64 on their CI, nor does support ppc64be + Linux, from the stack trace it looks like a V8 support issue and the fix probably won't be in Node.js.
